File: app/src/main/java/com/starostinvlad/fan/VideoScreen/PlayerModel/SerialPlayer.java
package com.starostinvlad.fan.VideoScreen.PlayerModel;

import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import android.util.Log;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.nodes.TextNode;
import org.jsoup.select.Elements;

import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

import io.reactivex.Maybe;
import io.reactivex.Observable;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class SerialPlayer {
    private final String TAG = getClass().getSimpleName();
    private String referer = "https://seriahd.tv";
    private String hash;

    private String DOMAIN = "";
    private OkHttpClient CLIENT;
    private String playerHost = "";
    private String title = "";

    public List<String> getInfoList() {
        return infoList;
    }

    public void setInfoList(List<String> infoList) {
        this.infoList = infoList;
    }

    private List<String> infoList = new ArrayList<>();

    public List<String> getReleaseDates() {
        return releaseDates;
    }

    public void setReleaseDates(List<String> releaseDates) {
        this.releaseDates = releaseDates;
    }

    private List<String> releaseDates = new ArrayList<>();


    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    private String description = "";

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public boolean isSubscribed() {
        return subscribed;
    }

    public void setSubscribed(boolean subscribed) {
        this.subscribed = subscribed;
    }

    private boolean subscribed;

    public SerialPlayer(String DOMAIN, OkHttpClient okHttpClient) {
        this.DOMAIN = DOMAIN;
        this.CLIENT = okHttpClient;
    }

    public Maybe<Hls> getHlsObject(final String url) {
        return Maybe.create(emitter -> {
            try {
                Document document = loadHlsFromUrl(url);
                Log.d(TAG, "getHlsObject: url: " + url);
                Log.d(TAG, "getHlsObject: doc: " + document);
                if (url.contains("responce.php")) {
                    emitter.onSuccess(parseHlsFromSecondPlayerPage(document));
                } else {
                    emitter.onSuccess(parseHlsFromDefualtPlayerPage(document));
                }
            } catch (Exception e) {
                emitter.onError(e);
            } finally {
                emitter.onComplete();
            }
        });
    }

    private Document loadHlsFromUrl(String url) throws IOException {
        Log.d(TAG, "getHlsUrl: iframe: " + url);
        Request getSeriaPage = new Request.Builder()
                .addHeader("referer", referer)
                .url(url)
                .get()
                .build();
        Response response = CLIENT.newCall(getSeriaPage).execute();

        if (response.code() == 200 & response.body() != null) {
            return Jsoup.parse(response.body().string());
        }
        throw new NullPointerException();
    }

    private Hls parseHlsFromSecondPlayerPage(Document document) {
        String playerJson = document.text();
        JsonObject jsonObject = JsonParser.parseString(playerJson).getAsJsonObject();
        String src = jsonObject.get("src").getAsString();
        String enSub = jsonObject.get("en_sub").getAsString();
        String ruSub = jsonObject.get("ru_sub").getAsString();
        return new Hls(src, enSub, ruSub);
    }

    private Hls parseHlsFromDefualtPlayerPage(Document document) {
        Element playerElement = document.selectFirst(".flowplayer");
        String playerJson = playerElement.attr("data-config");
        String src = JsonParser.parseString(playerJson).getAsJsonObject().get("hls").getAsString();
        String enSub = playerElement.attr("data-ru_subtitle");
        String ruSub = playerElement.attr("data-en_subtitle");
        return new Hls(src, enSub, ruSub);
    }

    private Document loadSerialPageFromUrl(String url) throws IOException {
        String currentUrl = url.contains(DOMAIN) ? url : DOMAIN + url;
        referer = currentUrl;
        Request getSeriaPage = new Request.Builder()
                .addHeader("referer", currentUrl)
                .url(currentUrl)
                .get()
                .build();
        Response response = CLIENT.newCall(getSeriaPage).execute();
        if (response.code() == 200 & response.body() != null) {
            return Jsoup.parse(
                    response
                            .body()
                            .string()
            );
        }
        return null;
    }

    public Observable<Serial> getSerial(String url) {
        return Observable.create(emitter -> {
            try {
                Document document = loadSerialPageFromUrl(url);
                hash = document.getElementsByAttributeValue("name", "user_hash").attr("value");

                title = document.getElementsByAttributeValue("property", "og:title").attr("content");

                Log.d(TAG, "title: " + title);

                List<TextNode> descriptionNodes = document.selectFirst("div.fdesc.full-text.clearfix").textNodes();
                for (TextNode descriptionNode : descriptionNodes.subList(2, descriptionNodes.size())) {
                    description += descriptionNode.text();
                }
                Log.d(TAG, "description: " + description);
                Elements serialInfoElements = document.select("div.sd-line");
                for (Element serialInfoElement : serialInfoElements) {
                    infoList.add(serialInfoElement.text());
                }
                Log.d(TAG, "serialInfo: " + infoList.toString());

                Elements releaseDateElements = document.select("tr.epscape_tr");
                for (Element date : releaseDateElements) {
                    releaseDates.add(date.text());
                }
                Log.d(TAG, "releaseDates: " + releaseDates.toString());

                subscribed = document.select(".fa.fa-star").hasClass("fav-added");
//                \"scode_begin\":\"([^}]*?dew.*?)\"}
                Pattern pattern = Pattern.compile("\"scode_begin\":\"([^}]*?dew.*?)\"");

                Matcher matcher = pattern.matcher(document.html());
                Serial serial;
                if (matcher.find()) {
                    String result = matcher.group(1);
                    result = result.replace("ifr:", "http").replace("\\", "");
                    Log.d(TAG, "getSerial: EXIST SECOND PLAYER! " + result);
                    serial = parseSerialFromSecondPlayer(getSecondPlayerPageFromUrl(result));
                    emitter.onNext(serial);
                } else {
                    serial = parseSerialFromDefautPlayer(document);
                    emitter.onNext(serial);
//                    throw new Exception("Данный сериал не поддерживается в приложении!");
                }
            } catch (Exception e) {
                emitter.onError(e);
            } finally {
                emitter.onComplete();
            }
        });
    }
    // TODO: 12.04.2021 разделить плеер и страницу сериала на две модели
    // TODO: 12.04.2021 оформить правильный нейминг 
    // TODO: 12.04.2021 собрать все файлы для работы с сайтом в отдельный пакет бизнес логики
    // TODO: 12.04.2021 попробовать переоформить структуру сериала(сезон->серия->озвучка)

    private Document getSecondPlayerPageFromUrl(String url) throws IOException {
        Request getSeriaPage = new Request.Builder()
                .addHeader("referer", referer)
                .url(url)
                .get()
                .build();
        Response response = CLIENT.newCall(getSeriaPage).execute();
        playerHost = Uri.parse(url).getHost();
        Document document = null;
        if (response.code() == 200 & response.body() != null) {
            document = Jsoup.parse(response.body().string());
        }
        return document;
    }

    private Serial parseSerialFromSecondPlayer(Document document) {
        Element inputData = document.selectFirst("#inputData");
        Serial serial = null;
        if (inputData != null && !inputData.text().isEmpty()) {
//                Log.d(TAG, "getHlsUrl: inputData: " + inputData.text());
            int id = Integer.parseInt(inputData.attr("data-playlist"));
            Gson gson = new GsonBuilder()
                    .setPrettyPrinting()
                    .setLenient()
                    .registerTypeAdapter(Serial.class, new SecondPlayerDataDeserializer(id, playerHost))
                    .create();
            serial = gson.fromJson(inputData.text(), Serial.class);
        }
        return serial;
    }

    private Serial parseSerialFromDefautPlayer(Document document) {
        String playerJson = document.select("script:nth-child(3)").html();
        Pattern pattern = Pattern.compile("init\\((.*?)\\);");

        Matcher matcher = pattern.matcher(playerJson);
        Serial serial = null;
        if (matcher.find()) {
            String result = matcher.group(1);
            result = "[" + result + "]";

            String path = Environment.getExternalStorageDirectory() + "/player.json";
            Log.d(TAG, "parseSerialFromDefautPlayer: path: " + path);
            try (FileOutputStream fileOutputStream = new FileOutputStream(path)) {
                byte[] arr = result.getBytes();
                fileOutputStream.write(arr);
            } catch (IOException e) {
                e.printStackTrace();
            } finally {
                Log.d(TAG, "deserialize: file writed!");
            }

            Log.d(TAG, "getPlayerList: " + result.substring(result.length() - 50));
            Gson gson = new GsonBuilder()
                    .setPrettyPrinting()
                    .setLenient()
                    .registerTypeAdapter(Serial.class, new PlayerDataDeserializer())
                    .create();
            serial = gson.fromJson(result, Serial.class);
            Log.d(TAG, "getSerial: serial: " + serial.toString());
        }
        return serial;
    }

    public Maybe<String> subscribeRequest(String id, boolean check) {
        return Maybe.create(emitter -> {
            String action = check ? "plus" : "minus";
            Request request = new Request
                    .Builder()
                    .url(DOMAIN + "/engine/ajax/controller.php?mod=favorites&fav_id=" + id + "&action=" + action + "&skin=seriahd&alert=1&user_hash=" + hash)
                    .get()
                    .build();
            try {
                Response response = CLIENT.newCall(request).execute();
                if (response.code() == 200 & response.body() != null) {
                    Document doc = Jsoup.parse(response.body().string());
                    emitter.onSuccess(doc.body().html());
                }
            } catch (IOException e) {
                emitter.onError(e);
            } finally {
                emitter.onComplete();
            }
        });
    }
}

File: app/src/main/java/com/starostinvlad/fan/VideoScreen/PlayerModel/SerialPlayer.kt
package com.starostinvlad.fan.VideoScreen.PlayerModel

import android.net.Uri
import android.os.Environment
import android.util.Log
import com.google.gson.GsonBuilder
import com.google.gson.JsonParser
import io.reactivex.Maybe
import io.reactivex.MaybeEmitter
import okhttp3.OkHttpClient
import okhttp3.Request
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import java.io.FileOutputStream
import java.io.IOException
import java.util.*
import java.util.Collections.emptyList
import java.util.regex.Pattern
import kotlin.Throws

class SerialPlayer(DOMAIN: String, okHttpClient: OkHttpClient) {
    private val TAG: String = this::class.simpleName!!
    private var referer: String = "https://seriahd.tv"
    private var hash = ""
    private var DOMAIN: String = ""
    private val CLIENT: OkHttpClient?
    private var playerHost: String? = ""
    var title = ""
    var description = ""
    var releaseDates: MutableList<String> = mutableListOf()
    var infoList: MutableList<String> = mutableListOf()
    var isSubscribed = false
    private var pageId = ""
    fun getHlsObject(url: String?): Maybe<Hls?> {
        return Maybe.create { emitter: MaybeEmitter<Hls?> ->
            try {
                val document = loadHlsFromUrl(url!!)
                Log.d(TAG, "getHlsObject: url: $url")
                Log.d(TAG, "getHlsObject: doc: $document")
                if (url.contains("responce.php")) {
                    emitter.onSuccess(parseHlsFromSecondPlayerPage(document))
                } else {
                    emitter.onSuccess(parseHlsFromDefualtPlayerPage(document))
                }
            } catch (e: Exception) {
                emitter.onError(e)
            } finally {
                emitter.onComplete()
            }
        }
    }

    @Throws(IOException::class)
    private fun loadHlsFromUrl(url: String): Document {
        Log.d(TAG, "getHlsUrl: iframe: $url")
        val getSeriaPage = Request.Builder()
                .addHeader("referer", referer)
                .url(url)
                .get()
                .build()
        val response = CLIENT!!.newCall(getSeriaPage).execute()
        if (response.code() == 200) {
            return Jsoup.parse(response.body()!!.string())
        }
        throw NullPointerException()
    }

    private fun parseHlsFromSecondPlayerPage(document: Document): Hls {
        val playerJson = document.text()
        val jsonObject = JsonParser.parseString(playerJson).asJsonObject
        val src = jsonObject["src"].asString
        val enSub = jsonObject["en_sub"].asString
        val ruSub = jsonObject["ru_sub"].asString
        return Hls(src, enSub, ruSub)
    }

    private fun parseHlsFromDefualtPlayerPage(document: Document): Hls {
        val playerElement = document.selectFirst(".flowplayer")
        val playerJson = playerElement.attr("data-config")
        val src = JsonParser.parseString(playerJson).asJsonObject["hls"].asString
        val enSub = playerElement.attr("data-ru_subtitle")
        val ruSub = playerElement.attr("data-en_subtitle")
        return Hls(src, enSub, ruSub)
    }

    @Throws(IOException::class)
    private fun loadSerialPageFromUrl(url: String): Document? {
        val currentUrl = if (url.contains(DOMAIN)) url else DOMAIN + url
        Log.d(TAG, "domain: $DOMAIN url : $currentUrl")
        referer = currentUrl
        val getSeriaPage = Request.Builder()
                .addHeader("referer", currentUrl)
                .url(currentUrl)
                .get()
                .build()
        val response = CLIENT!!.newCall(getSeriaPage).execute()
        return if (response.code() == 200) {
            Jsoup.parse(
                    response
                            .body()!!
                            .string()
            )
        } else null
    }

    fun getSerial(url: String): Maybe<Serial?> {
        return Maybe.create { emitter: MaybeEmitter<Serial?> ->
            try {
                val document = loadSerialPageFromUrl(url)
                pageId = url.substring(url.lastIndexOf("/"))
                pageId = pageId.substring(1, pageId.indexOf("-"))
                Log.d(TAG, "loadSerialPageFromUrl: pageID: $pageId")
                hash = document!!.getElementsByAttributeValue("name", "user_hash").attr("value")
                title = document.getElementsByAttributeValue("property", "og:title").attr("content")
                Log.d(TAG, "title: $title")
                val descriptionNodes = document.selectFirst("div.fdesc.full-text.clearfix").textNodes()
                for (descriptionNode in descriptionNodes.subList(2, descriptionNodes.size)) {
                    description += descriptionNode.text()
                }
                Log.d(TAG, "description: $description")
                val serialInfoElements = document.select("div.sd-line")
                for (serialInfoElement in serialInfoElements) {
                    infoList.add(serialInfoElement.text())
                }
                Log.d(TAG, "serialInfo: $infoList")
                val releaseDateElements = document.select("tr.epscape_tr")
                for (date in releaseDateElements) {
                    releaseDates.add(date.text())
                }
                Log.d(TAG, "releaseDates: $releaseDates")
                isSubscribed = document.select(".fa.fa-star").hasClass("fav-added")
                //                \"scode_begin\":\"([^}]*?dew.*?)\"}
                val pattern = Pattern.compile("\"scode_begin\":\"([^}]*?dew.*?)\"")
                val matcher = pattern.matcher(document.html())
                val serial: Serial?
                if (matcher.find()) {
                    var result = matcher.group(1)
                    result = result.replace("ifr:", "http").replace("\\", "")
                    Log.d(TAG, "getSerial: EXIST SECOND PLAYER! $result")
                    serial = parseSerialFromSecondPlayer(getSecondPlayerPageFromUrl(result))
                    serial?.id = pageId
                } else {
                    serial = parseSerialFromDefautPlayer(document)
                    serial?.id = pageId
                    //                    throw new Exception("Данный сериал не поддерживается в приложении!");
                }
                emitter.onSuccess(serial!!)
            } catch (e: Exception) {
                emitter.onError(e)
            } finally {
                emitter.onComplete()
            }
        }
    }

    // TODO: 12.04.2021 оформить правильный нейминг
    // TODO: 12.04.2021 собрать все файлы для работы с сайтом в отдельный пакет бизнес логики
    @Throws(IOException::class)
    private fun getSecondPlayerPageFromUrl(url: String): Document? {
        val getSeriaPage = Request.Builder()
                .addHeader("referer", referer)
                .url(url)
                .get()
                .build()
        val response = CLIENT!!.newCall(getSeriaPage).execute()
        playerHost = Uri.parse(url).host
        var document: Document? = null
        if (response.code() == 200) {
            document = Jsoup.parse(response.body()!!.string())
        }
        return document
    }

    private fun parseSerialFromSecondPlayer(document: Document?): Serial? {
        val inputData = document!!.selectFirst("#inputData")
        var serial: Serial? = null
        if (inputData != null && !inputData.text().isEmpty()) {
//                Log.d(TAG, "getHlsUrl: inputData: " + inputData.text());
            val id: Int = inputData.attr("data-playlist").toInt()
            val gson = GsonBuilder()
                    .setPrettyPrinting()
                    .setLenient()
                    .registerTypeAdapter(Serial::class.java, SecondPlayerDataDeserializer(id, playerHost))
                    .create()
            serial = gson.fromJson(inputData.text(), Serial::class.java)
        }
        return serial
    }

    private fun parseSerialFromDefautPlayer(document: Document?): Serial? {
        val playerJson = document!!.select("script:nth-child(3)").html()
        val pattern = Pattern.compile("init\\((.*?)\\);")
        val matcher = pattern.matcher(playerJson)
        var serial: Serial? = null
        if (matcher.find()) {
            var result = matcher.group(1)
            result = "[$result]"
            val path = Environment.getExternalStorageDirectory().toString() + "/player.json"
            Log.d(TAG, "parseSerialFromDefautPlayer: path: $path")
            try {
                FileOutputStream(path).use { fileOutputStream ->
                    val arr: ByteArray = result.toByteArray()
                    fileOutputStream.write(arr)
                }
            } catch (e: IOException) {
                e.printStackTrace()
            } finally {
                Log.d(TAG, "deserialize: file writed!")
            }
            Log.d(TAG, "getPlayerList: " + result.substring(result.length - 50))
            val gson = GsonBuilder()
                    .setPrettyPrinting()
                    .setLenient()
                    .registerTypeAdapter(Serial::class.java, PlayerDataDeserializer())
                    .create()
            serial = gson.fromJson(result, Serial::class.java)
            Log.d(TAG, "getSerial: serial: $serial")
        }
        return serial
    }

    fun subscribeRequest(id: String?, check: Boolean): Maybe<String?> {
        return Maybe.create { emitter: MaybeEmitter<String?> ->
            val action = if (check) "plus" else "minus"
            val request = Request.Builder()
                    .url("$DOMAIN/engine/ajax/controller.php?mod=favorites&fav_id=$id&action=$action&skin=seriahd&alert=1&user_hash=$hash")
                    .get()
                    .build()
            try {
                val response = CLIENT!!.newCall(request).execute()
                if (response.code() == 200) {
                    val doc = Jsoup.parse(response.body()!!.string())
                    emitter.onSuccess(doc.body().html())
                }
            } catch (e: IOException) {
                emitter.onError(e)
            } finally {
                emitter.onComplete()
            }
        }
    }


    init {
        this.DOMAIN = DOMAIN
        CLIENT = okHttpClient
    }
}